macaroonbakery/httpbakery/client.py

```diff
@@ -4,65 +4,213 @@
-            visit_page = utils.visit_page_with_browser
-        if 'agent-login' in cookies.keys():
-            self._visit_page = _visit_page_for_agent(cookies, key)
+        kwargs.setdefault('auth', self.auth())
+        return requests.request(method=method, url=url, **kwargs)
+
+    def acquire_discharge(self, cav, payload):
+        ''' Request a discharge macaroon from the caveat location
+        as an HTTP URL.
+        @param cav Third party {pymacaroons.Caveat} to be discharged.
+        @param payload External caveat data {bytes}.
+        @return The acquired macaroon {bakery.Macaroon}
+        '''
+        resp = self._acquire_discharge_with_token(cav, payload, None)
+        # TODO Fabrice what is the other http response possible ??
+        if resp.status_code == 200:
+            return bakery.Macaroon.from_dict(resp.json().get('Macaroon'))
+        cause = Error.from_dict(resp.json())
+        if cause.code != ERR_INTERACTION_REQUIRED:
+            raise DischargeError(cause.message)
+        if cause.info is None:
+            raise DischargeError(
+                'interaction-required response with no info: {}'.format(resp.json())
+            )
+        loc = cav.location
+        if not loc.endswith('/'):
+            loc = loc + '/'
+        token, m = self._interact(loc, cause, payload)
+        if m is not None:
+            # We've acquired the macaroon directly via legacy interaction.
+            return m
+        # Try to acquire the discharge again, but this time with
+        # the token acquired by the interaction method.
+        resp = self._acquire_discharge_with_token(cav, payload, token)
+        if resp.status_code == 200:
+            return bakery.Macaroon.deserialize_json(
+                resp.json().get('Macaroon'))
         else:
-            self._visit_page = visit_page
-        self._jar = cookies
-        self._key = key
+            raise DischargeError()
+
+    def _acquire_discharge_with_token(self, cav, payload, token):
+        req = {}
+        _add_json_binary_field(cav.caveat_id_bytes, req, 'id')
+        if token is not None:
+            _add_json_binary_field(token.value, req, 'token')
+            req['token-kind'] = token.kind
+        if payload is not None:
+            req['caveat64'] = base64.urlsafe_b64encode(payload).rstrip(
+                b'=').decode('utf-8')
+        target = relative_url(cav.location, 'discharge')
+        headers = {
+            BAKERY_PROTOCOL_HEADER: str(bakery.LATEST_BAKERY_VERSION)
+        }
+        return self.request('POST', target, data=req, headers=headers)
+
+    def _interact(self, location, error_info, payload):
+        '''Gathers a macaroon by directing the user to interact with a
+        web page. The error_info argument holds the interaction-required
+        error response.
+        @return DischargeToken, bakery.Macaroon
+        '''
+        if self._interaction_methods is None or len(self._interaction_methods) == 0:
```

Comment:
would it make sense to replace this with

if not self._interaction_methods:

?

Reply:
I *think* it's slightly nicer to use the explicit test, as it makes the expected type of self._interaction_methods clearer.